File: core/src/main/java/org/seppiko/wagashi/core/services/UserService.java
package org.seppiko.wagashi.core.services;

import org.seppiko.wagashi.core.mappers.RoleMapper;
import org.seppiko.wagashi.core.mappers.UserMapper;
import org.seppiko.wagashi.core.models.Role;
import org.seppiko.wagashi.core.models.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * @author Leonard Woo
 */
@Service
public class UserService implements UserDetailsService {

  @Autowired
  private UserMapper umapper;

  @Autowired
  private RoleMapper rmapper;

  @Override
  public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
    User user = umapper.queryByUsername(username);
    List<Role> roles = rmapper.queryByUserID(user.id());
    return new UserPrincipal(user, roles);
  }

}

File: core/src/main/kotlin/org/seppiko/wagashi/core/services/UserService.kt
/*
 * Copyright 2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.seppiko.wagashi.core.services

import org.seppiko.wagashi.core.mappers.UserMapper
import org.seppiko.wagashi.core.models.User
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

/**
 * @author Leonard Woo
 */
@Service
class UserService : UserDetailsService {

  @Autowired
  private lateinit var mapper: UserMapper

  @Throws(UsernameNotFoundException::class)
  override fun loadUserByUsername(username: String): UserDetails {
    val user: User? = mapper.queryByUsername(username)
    return UserPrincipal(user)
  }

}